contracts/treasury/contracts/src/ZrxTreasury.sol

```diff
@@ -165,43 +174,38 @@ contract ZrxTreasury is
         public
         override
     {
-        if (proposalId >= proposalCount()) {
-            revert("castVote/INVALID_PROPOSAL_ID");
-        }
-        if (hasVoted[proposalId][msg.sender]) {
-            revert("castVote/ALREADY_VOTED");
-        }
-
-        Proposal memory proposal = proposals[proposalId];
-        if (
-            proposal.voteEpoch != stakingProxy.currentEpoch() ||
-            _hasVoteEnded(proposal.voteEpoch)
-        ) {
-            revert("castVote/VOTING_IS_CLOSED");
-        }
-
-        uint256 votingPower = getVotingPower(msg.sender, operatedPoolIds);
-        if (votingPower == 0) {
-            revert("castVote/NO_VOTING_POWER");
-        }
+        return _castVote(msg.sender, proposalId, support, operatedPoolIds);
+    }
 
-        if (support) {
-            proposals[proposalId].votesFor = proposals[proposalId].votesFor
-                .safeAdd(votingPower);
-            hasVoted[proposalId][msg.sender] = true;
-        } else {
-            proposals[proposalId].votesAgainst = proposals[proposalId].votesAgainst
-                .safeAdd(votingPower);
-            hasVoted[proposalId][msg.sender] = true;
-        }
+    /// @dev Casts a vote for the given proposal, by signature.
+    ///      Only callable during the voting period for that proposal.
+    ///      See `getVotingPower` for how voting power is computed.
+    /// @param proposalId The ID of the proposal to vote on.
+    /// @param support Whether to support the proposal or not.
+    /// @param operatedPoolIds The pools operated by `msg.sender`. The
+    ///        ZRX currently delegated to those pools will be accounted
+    ///        for in the voting power.
+    /// @param v the v field of the signature
+    /// @param r the r field of the signature
+    /// @param s the s field of the signature
+    function castVoteBySignature(
+        uint256 proposalId,
+        bool support,
+        bytes32[] memory operatedPoolIds,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    )
+        public
+        override
+    {
+        bytes32 domainSeparator = keccak256(abi.encode(DOMAIN_TYPEHASH, keccak256(bytes(name)), _getChainId(), address(this)));
```

Comment:
`domainnSeparator` is basically constant and a good candidate for an [`immutable`](https://docs.soliditylang.org/en/v0.8.6/contracts.html?highlight=immutable#immutable) set once in the constructor.

Reply:
This is how it is done elsewhere in the codebase: https://github.com/0xProject/protocol/blob/24397c51a8c7bf704948c8fc6874843bccd5d244/contracts/zero-ex/contracts/src/fixins/FixinEIP712.sol#L35

Reply:
Should I directly use the domain separator from `FixinEIP712`? @Recmo

Reply:
No, it's better to create unique domain separators for each distinct use of signatures. The domain separators are there to make sure a signature is not accidentally valid for two different use-cases.